maven-plugin/src/main/java/org/sonatype/nexus/maven/staging/StagingDeployMojo.java

```diff
@@ -106,12 +113,47 @@ private void doExecute() throws MojoFailureException, MojoExecutionException {
     catch (MojoExecutionException e) {
       throw e;
     }
+    catch (RepositoryManagerException ex) {
+        Optional<String> message = findJsonErrorMessage(ex);
+        if(message.isPresent())
+        	throw new MojoFailureException(ex.getMessage() + ": " + message.get(), ex);
+    	throw new MojoFailureException(ex.getMessage(), ex);
+    }
     catch (Exception ex) {
       throw new MojoFailureException(ex.getMessage(), ex);
     }
   }
 
-  private String getProvidedOrGeneratedTag() {
+  private Optional<String> findJsonErrorMessage(Throwable e) {
+	if(e == null)
+	  return Optional.empty();
+
+	if(e instanceof HttpResponseException) {
+		Method[] declaredMethods = e.getClass().getDeclaredMethods();
+		for (Method method : declaredMethods) {
+			if(method.getParameterCount() == 0 && isPublic(method.getModifiers()) && Optional.class.isAssignableFrom(method.getReturnType())) {
+				try {
+					Optional<?> result = (Optional<?>) method.invoke(e);
+					if(result.isPresent()) {
+						Object potentialMessage = result.get();
+						// filter out the raw json message
+						if(potentialMessage instanceof String && !((String)potentialMessage).contains("\"message\"")) {
+							return Optional.of((String)potentialMessage);
+						}
+					}
+				} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e1) {
+					// ignore and continue
+				}
+			}
+		}
+		// nothing found
+        return Optional.empty();
+	}
+
+	return findJsonErrorMessage(e.getCause());
```

Comment:
Please add tests to exercise all parts of the recursive flow in this new method.

Reply:
@bobotimi I think it would be better if the team introduces a proper API to make the error response from Nexus part of the Nexus client API. For some reason the code is obfuscated, which doesn't make a lot of sense.

The reflection shouldn't be used as a long term solution.